File: backend/src/model/Delivery.ts
import { Document, model, Model, Schema } from 'mongoose'
import { IDelivery } from '../types/delivery'

export enum DeliveryStatus {
  UNASSIGNED = 'UNASSIGNED',
  PENDING = 'PENDING',
  ASSIGNED = 'ASSIGNED',
  IN_TRANSIT = 'IN_TRANSIT',
  DELIVERED = 'DELIVERED'
}

export interface DeliveryDoc extends IDelivery, Document {}

interface DeliveryModel extends Model<DeliveryDoc> {
  build: (attr: IDelivery) => DeliveryDoc
}

const locationSchema = new Schema({
  coordinate: {
    longitude: {
      type: Number,
      required: true
    },
    latitude: {
      type: Number,
      required: true
    }
  },
  country: {
    type: String,
    required: true
  },
  postalCode: {
    type: Number,
    required: true
  },
  city: {
    type: String,
    required: true
  },
  address: {
    type: String,
    required: true
  }
})

const deliverySchema = new Schema({
  title: {
    type: String,
    required: true
  },
  description: {
    type: String,
    required: true
  },
  sourceLocation: {
    type: locationSchema,
    required: true
  },
  destinationLocation: {
    type: locationSchema,
    required: true
  },
  clientUser: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  transporterUser: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: false
  },
  pickUpFrom: {
    type: Date,
    required: true
  },
  pickUpUntil: {
    type: Date,
    required: true
  },
  price: {
    type: Number,
    required: true
  },
  clientRating: {
    type: Number,
    required: false
  },
  transporterRating: {
    type: Number,
    required: false
  },
  capacity: {
    weight: {
      type: Number,
      required: true
    },
    height: {
      type: Number,
      required: true
    },
    length: {
      type: Number,
      required: true
    },
    width: {
      type: Number,
      required: true
    }
  },
  status: {
    type: String,
    enum: [
      DeliveryStatus.UNASSIGNED,
      DeliveryStatus.PENDING,
      DeliveryStatus.IN_TRANSIT,
      DeliveryStatus.DELIVERED,
      DeliveryStatus.ASSIGNED
    ],
    default: DeliveryStatus.UNASSIGNED,
    required: true
  },
  pictureUrl: {
    type: String,
    required: false
  }
})
deliverySchema.set('timestamps', true)
deliverySchema.statics.build = (attr: IDelivery) => new Delivery(attr)

const Delivery = model<DeliveryDoc, DeliveryModel>('Delivery', deliverySchema)

export { Delivery }

File: android/customer/app/src/main/java/hu/bme/aut/deliveryappforcustomers/model/Delivery.kt
package hu.bme.aut.android.deliveryapp.model

import hu.bme.aut.deliveryappforcustomers.model.Capacity
import hu.bme.aut.deliveryappforcustomers.model.DeliveryStatus
import hu.bme.aut.deliveryappforcustomers.model.Location
import java.io.Serializable

data class Delivery(
    val clientUser: String,
    val transporterUser: String? = null,
    val pictureUrl: String? = null,
    val description: String,
    val destinationLocation: Location,
    val title: String,
    val clientRating: Float? = null,
    val pickUpUntil: String,
    val price: Float,
    val sourceLocation: Location,
    val transporterRating: Float? = null,
    val createdAt: String,
    val _id: String? = null,
    val pickUpFrom: String,
    var status: DeliveryStatus? = null,
    val updatedAt: String? = null,
    val capacity: Capacity? = null,
) : Serializable
